### PR TITLE
Added GetDLRDeviceType api

### PR DESCRIPTION
**Pull request by cnv1989**
Added APIs to expose DeviceType from metadata of compilation.

Thanks for contributing to DLR! By submitting this pull request, you confirm that your contribution is made under the terms of the [Apache 2.0 license](https://github.com/neo-ai/neo-ai-dlr/blob/master/LICENSE).

Please refer to our [guideline](https://github.com/neo-ai/neo-ai-dlr/blob/master/CONTRIBUTING.md) for useful information and tips.

### CODE REVIEW

File: include/dlr_hexagon/dlr_hexagon.h [hunk 1]
```diff
@@ -8,6 +8,7 @@ namespace dlr {
 /*! \brief Get the paths of the Hexagon model files.
  */
 std::string GetHexagonModelFile(const std::string& dirname);
+std::string GetHexagonModelFile(const std::string& dirname);
```
Comment (trevor-m) [inline include/dlr_hexagon/dlr_hexagon.h:11]: Duplicated line?

File: src/dlr_hexagon/dlr_hexagon.cc [hunk 2]
```diff
@@ -48,6 +48,20 @@ bool dlr::FindHexagonNNSkelFile(const std::string& dirname) {
...
+std::string dlr::GetMetadataFile(const std::string& dirname) {
+  // Scan Dir to find libhexagon_nn_skel.so
+  std::vector<std::string> paths_vec;
+  ListDir(dirname, paths_vec);
+  for (auto filename : paths_vec) {
+    if (EndsWith(filename, ".meta")) {
+      return filename;
+    }
+  }
+  LOG(INFO) << "libhexagon_nn_skel.so file is not found under folder: "
```
Comment (trevor-m) [inline src/dlr_hexagon/dlr_hexagon.cc:60]: This message looks out of place

File: src/dlr_common.cc [hunk 3]
```diff
@@ -106,3 +106,40 @@ const std::vector<int64_t>& DLRModel::GetInputShape(int index) const {
...
+  return !this->metadata_.is_null(); 
+}
+
+
+const DLDeviceType DLRModel::GetDeviceTypeFromMetadata() const {
+  if (!this->HasMetadata()) {
+    throw dmlc::Error("No metadata file was found!");
+  }
+  try {
+    const std::string &target_backend = metadata_.at("Requirements")
```
Comment (trevor-m) [inline src/dlr_common.cc:121]: `const std::string &target_backend` -> `const std::string& target_backend`

**Review by trevor-m — CHANGES_REQUESTED** [review]
Thanks @cnv1989 ! Could you also add a check so that if the device type is available from the metadata but the user provides the wrong device type we give an error message? Somewhere in `SetupTVMModule`, `SetupTreeliteModule`, etc. Maybe make a helper function in DLRModel base class like `VerifyDeviceTypeIfAvailable()`.

File: src/dlr_common.cc [hunk 4]
```diff
@@ -106,3 +106,40 @@ const std::vector<int64_t>& DLRModel::GetInputShape(int index) const {
...
+}
+
+const DLDeviceType dlr::GetDeviceTypeForTargetBackend(const std::string& target_backend) {
+  if (target_backend == "llvm") {
+    return DLDeviceType::kDLCPU;
+  } else if (target_backend == "cuda") {
+    return DLDeviceType::kDLGPU;
+  } else if (target_backend == "arm_cpu") {
+    return DLDeviceType::kDLCPU;
+  } else if (target_backend == "mali") {
```
Comment (trevor-m) [inline src/dlr_common.cc:139]: Mali should be `kDLOpenCL`

**Review by trevor-m — CHANGES_REQUESTED** [review]

**Review by trevor-m — APPROVED** [review]

**Comment by apivovarov** [comment]
WITH_HEXAGON build is broken https://github.com/neo-ai/neo-ai-dlr/issues/251

**Pull request merged**